neo/Ledger/Blockchain.cs

```diff
@@ -0,0 +1,705 @@
+            return RelayResultReason.Succeed;
+        }
+
+        private void OnNewHeaders(Header[] headers)
+        {
+            using (Snapshot snapshot = GetSnapshot())
+            {
+                foreach (Header header in headers)
+                {
+                    if (header.Index - 1 >= header_index.Count) break;
+                    if (header.Index < header_index.Count) continue;
+                    if (!header.Verify(snapshot)) break;
+                    header_index.Add(header.Hash);
+                    snapshot.Blocks.Add(header.Hash, new BlockState
+                    {
+                        SystemFeeAmount = 0,
+                        TrimmedBlock = header.Trim()
+                    });
+                    snapshot.HeaderHashIndex.GetAndChange().Hash = header.Hash;
+                    snapshot.HeaderHashIndex.GetAndChange().Index = header.Index;
+                }
+                SaveHeaderHashList(snapshot);
+                snapshot.Commit();
+            }
+            UpdateCurrentSnapshot();
+            system.TaskManager.Tell(new TaskManager.HeaderTaskCompleted(), Sender);
+        }
+
+        private RelayResultReason OnNewTransaction(Transaction transaction)
+        {
+            const int MemoryPoolSize = 50000;
+            if (transaction.Type == TransactionType.MinerTransaction)
+                return RelayResultReason.Invalid;
+            if (ContainsTransaction(transaction.Hash))
+                return RelayResultReason.AlreadyExists;
+            if (!transaction.Verify(currentSnapshot, mem_pool.Values))
+                return RelayResultReason.Invalid;
+            mem_pool.TryAdd(transaction.Hash, transaction);
+            if (mem_pool.Count > MemoryPoolSize)
+            {
+                UInt256[] delete = mem_pool.Values.AsParallel()
+                    .OrderBy(p => p.NetworkFee / p.Size)
+                    .ThenBy(p => p.NetworkFee)
+                    .ThenBy(p => new BigInteger(p.Hash.ToArray()))
+                    .Take(mem_pool.Count - MemoryPoolSize)
+                    .Select(p => p.Hash)
+                    .ToArray();
+                foreach (UInt256 hash in delete)
+                    mem_pool.TryRemove(hash, out _);
+            }
+            if (!mem_pool.ContainsKey(transaction.Hash))
+                return RelayResultReason.OutOfMemory;
+            system.LocalNode.Tell(new LocalNode.RelayDirectly { Inventory = transaction });
+            return RelayResultReason.Succeed;
+        }
+
+        private void OnPersistCompleted(Block block)
+        {
+            block_cache.Remove(block.Hash);
+            foreach (Transaction tx in block.Transactions)
+                mem_pool.TryRemove(tx.Hash, out _);
+            Transaction[] remain = mem_pool.Values.ToArray();
+            mem_pool.Clear();
+            foreach (Transaction tx in remain)
+            {
+                if (!tx.Verify(currentSnapshot, mem_pool.Values)) continue;
```

Comment:
This code should not reverify all the transactions after each block persisted if it is currently syncing the chain.
It needs to keep some sort of temporary pool of transactions or something like was done previously so that it can leave them unverified if receiving blocks at a high rate due to syncing the chain.

The current 2.8.0 code has this feature. Let's not lose the ability to sync quickly.

Reply:
Also, in the case that transaction verification is taking longer than the ideal block time it should stop verifying the transactions and leave the rest in the temporary pool, so it doesn't potentially delay consensus if it is a consensus node and there is a large mem_pool.

Reply:
I updated the code, and put unverified transactions in the memory pool back into the message queue after the block is persisted.

Reply:
Ok, so that puts them all back in the message queue, but then they all have to be processed before it can process the next block, right? Won’t that still cause very slow performance of syncing the chain with a large mem_pool?

Reply:
https://github.com/neo-project/neo/blob/e53235c3be45512bcb5f761494c7dc20174ac78d/neo/Ledger/Blockchain.cs#L687-L699

The priority of `Block` message is higher than `Transaction` message, so if there is a new block incoming, the block will be processed first.

Reply:
Why incur the extra overhead of creating an object per transaction plus an enqueue and dequeue of each item if they are just going to be processed sequentially right away anyway?

Reply:
Ah, ok the higher priority makes it work then. Sounds good.

Reply:
Ah, ok the higher priority makes it work then. Sounds good.

Reply:
Ah, ok the higher priority makes it work then. Sounds good.